### PR TITLE
Dragan

### PR DESCRIPTION
**Pull request by martyn**
https://arxiv.org/pdf/1705.07215.pdf

Implements with an optional new `loss` parameter.  Example snippet
```json
{
"loss":{
  "gradient_penalty": 10,
  "gradient_penalty_type": "dragan" 
}
}
```

Technically works with all loss functions

https://www.youtube.com/watch?v=DQWS9qSZH2Q&feature=youtu.be

### CODE REVIEW

File: hypergan/losses/base_loss.py [hunk 1]
```diff
@@ -93,19 +93,21 @@ def half(tens, second):
...
         g = generator.sample
         discriminator = self.discriminator or gan.discriminator
         shape = [1 for t in g.get_shape()]
         shape[0] = gan.batch_size()
         uniform_noise = tf.random_uniform(shape=shape,minval=0.,maxval=1.)
         print("[gradient penalty] applying x:", x, "g:", g, "noise:", uniform_noise)
-        interpolates = x + uniform_noise * (g - x)
+        if config.gradient_penalty_type == 'dragan':
+            mean, variance = tf.nn.moments(x, axes=[0, 1, 2, 3])
+            interpolates = x + uniform_noise * 0.5 * variance * tf.random_uniform(shape=ops.shape(x), minval=0.,maxval=1.)
```
Comment (mikkel) [inline hypergan/losses/base_loss.py]: A couple of open questions:
* Should we bound this to [-1, 1] ?
* This equation is `uniform_noise` (values 0-1) * `variance` * 0.5 * `alpha`(0-1) + x.  Should one of those be values -1 to 1?

We may want to try gaussians, maybe even do some fun stuff with the tensorflow image processing utils https://www.tensorflow.org/api_guides/python/image  (Image Adjustments).

We can also render `interpolates` in a sampler.  I wonder if `nearby_image` would be a better name.

File: hypergan/losses/base_loss.py [hunk 2]
```diff
@@ -90,22 +96,42 @@ def half(tens, second):
...
         shape[0] = gan.batch_size()
         uniform_noise = tf.random_uniform(shape=shape,minval=0.,maxval=1.)
         print("[gradient penalty] applying x:", x, "g:", g, "noise:", uniform_noise)
-        interpolates = x + uniform_noise * (g - x)
+        if config.gradient_penalty_type == 'dragan':
+            axes = [0, 1, 2, 3]
+            if len(ops.shape(x)) == 2:
+                axes = [0, 1]
+            mean, variance = tf.nn.moments(x, axes=axes)
+            interpolates = x + uniform_noise * 0.5 * variance * tf.random_uniform(shape=ops.shape(x), minval=0.,maxval=1.)
```
Comment (mikkel) [inline hypergan/losses/base_loss.py]: A couple of open questions:
* Should we bound this to [-1, 1] ?
* This equation is `uniform_noise` (values 0-1) * `variance` * 0.5 * `alpha`(0-1) + x.  Should `alpha` or `uniform_noise` be values -1 to 1?

We may want to try gaussians, maybe even do some fun stuff with the tensorflow image processing utils https://www.tensorflow.org/api_guides/python/image  (Image Adjustments).

We can also render `interpolates` in a sampler.  I wonder if `nearby_image` would be a better name.

Comment (martyn) [inline hypergan/losses/base_loss.py]: Maybe we should treat the second random uniform as an encoder that defaults to the paper.
